src/core/lib/transport/metadata.h

```diff
@@ -136,10 +136,15 @@ bool grpc_mdelem_eq(grpc_mdelem a, grpc_mdelem b);
  * static, and that the keys match. This most commonly happens when processing
  * metadata batch callouts in initial/trailing filters. In this case, fastpath
  * grpc_mdelem_eq and remove unnecessary checks. */
-bool grpc_slice_eq_static(const grpc_slice& a, const grpc_slice& b_static);
+int grpc_slice_differs_slowpath(const grpc_slice& a,
+                                const grpc_slice& b_not_inline);
 inline bool grpc_mdelem_static_value_eq(grpc_mdelem a, grpc_mdelem b_static) {
   if (a.payload == b_static.payload) return true;
-  return grpc_slice_eq_static(GRPC_MDVALUE(a), GRPC_MDVALUE(b_static));
+  const grpc_slice& a_slice = GRPC_MDVALUE(a);
+  const grpc_slice& b_slice = GRPC_MDVALUE(b_static);
+  return a_slice.refcount == b_slice.refcount
```

Comment:
Could we just call grpc_slice_eq_static here?

Reply:
So this will also tie into a comment that Mark made, regarding why I'm doing a forward declaration as opposed to a an include of slice_internal.h.

Basically, there is a dependency cycle. slice_internal.h includes static_metadata.h which includes metadata.h. So this means we cannot include slice_internal.h here. So we have to forward declare.

Since grpc_slice_eq_static is inlined in its definition in slice_internal.h, just forward declaring means that we'll get a linker error with just a forward declaration.

However, the slow path is not declared inline, so my terrible hack here is to just duplicate a bit of implementation and then call the slow path method which won't throw a linker error.

TLDR: I don't call it because it would fail during the link process, and this is a hack.

Reply:
Is there some other way to solve it so we don't have to use the hack? Could we move grpc_slice_eq_static to a different file, then have metadata.h include that file?

Reply:
That is an option, but it's not clear where it should go. I am open to suggestions, since I am also not a huge fan of what I have done here.

Reply:
cc @markdroth

Reply:
Maybe make a new file (or files) called slice_utils.h (and maybe a slice_utils.cc)? And it could live in the same directory as slice_internal.h

Reply:
Sounds reasonable - added slice_utils.h and removed the forward declaration.